src/shims/thread.rs

```diff
@@ -0,0 +1,138 @@
+
+        this.tcx.sess.warn(
+            "thread support is experimental. \
+             For example, Miri does not detect data races yet.",
+        );
+
+        let new_thread_id = this.create_thread()?;
+        let old_thread_id = this.set_active_thread(new_thread_id)?;
+
+        let thread_info_place = this.deref_operand(thread)?;
+        let thread_info_type = thread.layout.ty
+            .builtin_deref(true)
+            .ok_or_else(|| err_ub_format!(
+                "wrong signature used for `pthread_create`: first argument must be a raw pointer."
+            ))?
+            .ty;
+        let thread_info_layout = this.layout_of(thread_info_type)?;
+        this.write_scalar(
+            Scalar::from_uint(new_thread_id.index() as u128, thread_info_layout.size),
+            thread_info_place.into(),
+        )?;
+
+        let fn_ptr = this.read_scalar(start_routine)?.not_undef()?;
+        let instance = this.memory.get_fn(fn_ptr)?.as_instance()?;
+
+        let func_arg = this.read_immediate(arg)?;
+        let func_args = [*func_arg];
+
+        let ret_place =
+            this.allocate(this.layout_of(this.tcx.types.usize)?, MiriMemoryKind::Machine.into());
+
+        this.call_function(
+            instance,
+            &func_args[..],
+            Some(ret_place.into()),
+            StackPopCleanup::None { cleanup: true },
+        )?;
+
+        this.set_active_thread(old_thread_id)?;
+
+        Ok(0)
+    }
+
+    fn pthread_join(
+        &mut self,
+        thread: OpTy<'tcx, Tag>,
+        retval: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.is_null(this.read_scalar(retval)?.not_undef()?)? {
+            throw_unsup_format!("Miri supports pthread_join only with retval==NULL");
+        }
+
+        let thread_id = this.read_scalar(thread)?.not_undef()?.to_machine_usize(this)?;
+        this.join_thread(thread_id.into())?;
+
+        Ok(0)
+    }
+
+    fn pthread_detach(&mut self, thread: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let thread_id = this.read_scalar(thread)?.not_undef()?.to_machine_usize(this)?;
+        this.detach_thread(thread_id.into())?;
+
+        Ok(0)
+    }
+
+    fn pthread_self(&mut self, dest: PlaceTy<'tcx, Tag>) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        let thread_id = this.get_active_thread()?;
+        this.write_scalar(Scalar::from_uint(thread_id.index() as u128, dest.layout.size), dest)
+    }
+
+    fn prctl(
+        &mut self,
+        option: OpTy<'tcx, Tag>,
+        arg2: OpTy<'tcx, Tag>,
+        arg3: OpTy<'tcx, Tag>,
+        arg4: OpTy<'tcx, Tag>,
+        arg5: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        // prctl last 5 arguments are declared as variadic. Therefore, we need
+        // to check their types manually.
```

Comment:
I don't think we do this anywhere else, and also the fact taht they are variadic makes no difference.

I think this is basically https://github.com/rust-lang/miri/issues/1355, and we should find a proper solution for it -- once we do, `check_arg` here is just redundant. So I'd prefer if we would not add it.